From 046f5375514b4c1b4cf092e2f2c5bc854ee75a01 Mon Sep 17 00:00:00 2001
From: srinidhira0 <srinidhi.rao@broadcom.com>
Date: Fri, 19 Sep 2025 11:41:09 +0000
Subject: [PATCH] jitterentropy: Defer jent_mod_init in non fips boot

  -  In linux-esx, when fips is not enabled in command line, then the
     kernel is expected to boot as quickly as possible. when fips is 0
     it is not neccessary to perform conditioning on the 1024 TEST LOOP
     during jent_entropy_init. Optimize by skipping conditioning and
     also defer the jent_mod_init work to a delayed workqueue so that
     initialization can continue in parallel instead of blocking other
     initcalls.

Signed-off-by: srinidhira0 <srinidhi.rao@broadcom.com>
---
 jitterentropy-kcapi.c | 96 +++++++++++++++++++++++++++++++++++++++++--
 jitterentropy.c       | 28 +++++++++----
 jitterentropy.h       |  3 ++
 3 files changed, 115 insertions(+), 12 deletions(-)

diff --git a/jitterentropy-kcapi.c b/jitterentropy-kcapi.c
index dceadb7..a731dcb 100644
--- a/jitterentropy-kcapi.c
+++ b/jitterentropy-kcapi.c
@@ -46,6 +46,9 @@
 #include <linux/time.h>
 #include <crypto/internal/rng.h>
 
+#include <linux/jiffies.h>
+#include <linux/workqueue.h>
+
 #include "photon-jitterentropy-v6.12/jitterentropy.h"
 #include "photon-jitterentropy-v6.12/jitterentropy_canister_wrapper.h"
 
@@ -57,6 +60,17 @@ module_param_string(mem_access_type, mem_access_method,
                                         ARRAY_SIZE(mem_access_method), 0444);
 MODULE_PARM_DESC(mem_access_type, "jent Mem Block Access default=random");
 
+/***************************************************************************
+ * wait for jent init completion helpers
+ ***************************************************************************/
+
+static int sync_on_jent_init(void);
+static void jent_delayed_init(struct work_struct *work);
+
+static atomic_t jent_init_complete;
+static DECLARE_DELAYED_WORK(jent_init_work, jent_delayed_init);
+static DEFINE_MUTEX(jent_init_mutex);
+
 /***************************************************************************
  * Helper function
  ***************************************************************************/
@@ -200,6 +214,11 @@ static void jent_kcapi_cleanup(struct crypto_tfm *tfm)
 {
 	struct jitterentropy *rng = crypto_tfm_ctx(tfm);
 
+	if (!atomic_read(&jent_init_complete) && !jcw_fips_enabled()) {
+		cancel_delayed_work_sync(&jent_init_work);
+		pr_devel("Invoked %s, canceled sync\n", __func__);
+	}
+
 	spin_lock(&rng->jent_lock);
 
 	if (rng->sdesc) {
@@ -225,6 +244,12 @@ static int jent_kcapi_init(struct crypto_tfm *tfm)
 	struct shash_desc *sdesc;
 	int size, ret = 0;
 
+	/* If jent init was delayed, sync till jent init is complete */
+	if (!jcw_fips_enabled()) {
+		if (!sync_on_jent_init())
+			pr_devel("Invoked %s. Sync jent_init\n", __func__);
+	}
+
 	spin_lock_init(&rng->jent_lock);
 
 	/*
@@ -277,6 +302,9 @@ static int jent_kcapi_random(struct crypto_rng *tfm,
 	struct jitterentropy *rng = crypto_rng_ctx(tfm);
 	int ret = 0;
 
+	if (!sync_on_jent_init())
+		pr_devel("Invoked %s\n", __func__);
+
 	spin_lock(&rng->jent_lock);
 
 	ret = jent_read_entropy(rng->entropy_collector, rdata, dlen);
@@ -311,6 +339,9 @@ static int jent_kcapi_random(struct crypto_rng *tfm,
 static int jent_kcapi_reset(struct crypto_rng *tfm,
 			    const u8 *seed, unsigned int slen)
 {
+	if (!sync_on_jent_init())
+		pr_devel("Invoked %s\n", __func__);
+
 	return 0;
 }
 
@@ -329,13 +360,16 @@ static struct rng_alg jent_alg = {
 	}
 };
 
-static int __init jent_mod_init(void)
+static int invoke_jent_init(void)
 {
 	SHASH_DESC_ON_STACK(desc, tfm);
 	struct crypto_shash *tfm;
 	int ret = 0;
+	int init_in_progress = 0;
 
 	jent_testing_init();
+	if (!jcw_fips_enabled())
+		init_in_progress = JENT_INIT;
 
 	tfm = crypto_alloc_shash(JENT_CONDITIONING_HASH, 0, 0);
 	if (IS_ERR(tfm)) {
@@ -345,24 +379,78 @@ static int __init jent_mod_init(void)
 
 	desc->tfm = tfm;
 	crypto_shash_init(desc);
-	ret = jent_entropy_init(CONFIG_CRYPTO_JITTERENTROPY_OSR, 0, desc, NULL);
+	ret = jent_entropy_init(CONFIG_CRYPTO_JITTERENTROPY_OSR, init_in_progress, desc, NULL);
 	shash_desc_zero(desc);
 	crypto_free_shash(tfm);
-	if (ret) {
+	if (unlikely(ret)) {
 		/* Handle permanent health test error */
 		if (fips_enabled)
 			panic("jitterentropy: Initialization failed with host not compliant with requirements: %d\n", ret);
 
 		jent_testing_exit();
 		pr_info("jitterentropy: Initialization failed with host not compliant with requirements: %d\n", ret);
-		return -EFAULT;
+		ret = -EFAULT;
 	}
 	pr_info("jitterentropy: Initialized Jitterentropy\n");
+	return ret;
+}
+
+static int sync_on_jent_init(void)
+{
+	int ret = -EALREADY;
+
+	if (unlikely(mutex_lock_interruptible(&jent_init_mutex)))
+		return -EINTR;
+	if (!atomic_read(&jent_init_complete) && !jcw_fips_enabled()) {
+		if (cancel_delayed_work_sync(&jent_init_work)) {
+			ret = invoke_jent_init();
+			if (unlikely(ret)) {
+			WARN_ONCE(1, "jent: Init failed %d\n", ret);
+				crypto_unregister_rng(&jent_alg);
+				goto check_exit;
+			}
+		}
+		atomic_set(&jent_init_complete, true);
+		ret = 0;
+	}
+check_exit:
+	mutex_unlock(&jent_init_mutex);
+	return ret;
+}
+
+static void jent_delayed_init(struct work_struct *work)
+{
+	int ret = 0;
+
+	ret = invoke_jent_init();
+	if (unlikely(ret)) {
+		WARN_ONCE(1, "jitterentropy: Delayed Initialization Failure %d\n", ret);
+		crypto_unregister_rng(&jent_alg);
+	} else {
+		pr_devel("Jent delayed init completed successfully\n");
+		atomic_set(&jent_init_complete, true);
+	}
+}
+
+static int __init jent_mod_init(void)
+{
+	int ret = 0;
+
+	if (!jcw_fips_enabled()) {
+		schedule_delayed_work(&jent_init_work, HZ);
+	} else {
+		ret = invoke_jent_init();
+		if (unlikely(ret))
+			return ret;
+	}
 	return crypto_register_rng(&jent_alg);
 }
 
 static void __exit jent_mod_exit(void)
 {
+	if (!atomic_read(&jent_init_complete) && !jcw_fips_enabled())
+		cancel_delayed_work_sync(&jent_init_work);
+
 	jent_testing_exit();
 	crypto_unregister_rng(&jent_alg);
 }
diff --git a/jitterentropy.c b/jitterentropy.c
index ddfe1ed..d992a73 100644
--- a/jitterentropy.c
+++ b/jitterentropy.c
@@ -730,7 +730,8 @@ static void jent_memaccess_sequential(struct rand_data *ec, __u64 loop_cnt)
  *
  * @return result of stuck test
  */
-static int jent_measure_jitter(struct rand_data *ec, __u64 *ret_current_delta)
+static int jent_measure_jitter(struct rand_data *ec, __u64 *ret_current_delta,
+				bool init_in_progress)
 {
 	__u64 time = 0;
 	__u64 current_delta = 0;
@@ -761,8 +762,10 @@ static int jent_measure_jitter(struct rand_data *ec, __u64 *ret_current_delta)
 	stuck = jent_stuck(ec, current_delta);
 
 	/* Now call the next noise sources which also injects the data */
-	if (jent_condition_data(ec, current_delta, stuck))
-		stuck = 1;
+	/* Skip conditioning if fips_enabled is 0 & bootup is in progress */
+	if (init_in_progress == false)
+		if (jent_condition_data(ec, current_delta, stuck))
+			stuck = 1;
 
 	/* return the raw entropy value */
 	if (ret_current_delta)
@@ -785,11 +788,11 @@ static void jent_gen_entropy(struct rand_data *ec)
 		safety_factor = JENT_ENTROPY_SAFETY_FACTOR;
 
 	/* priming of the ->prev_time value */
-	jent_measure_jitter(ec, NULL);
+	jent_measure_jitter(ec, NULL, false);
 
 	while (!jent_health_failure(ec)) {
 		/* If a stuck measurement is received, repeat measurement */
-		if (jent_measure_jitter(ec, NULL))
+		if (jent_measure_jitter(ec, NULL, false))
 			continue;
 
 		/*
@@ -926,7 +929,12 @@ struct rand_data *jent_entropy_collector_alloc(unsigned int osr,
 	jent_lag_init(entropy_collector, osr);
 
 	/* fill the data pad with non-zero values */
-	jent_gen_entropy(entropy_collector);
+	/*
+	 * Optimize jent_init during bootup by skipping the call to
+	 * jent_gen_entropy when fips_enabled is 0 and init is in progress
+	 */
+	if (jcw_fips_enabled() || (flags != JENT_INIT))
+		jent_gen_entropy(entropy_collector);
 
 	return entropy_collector;
 }
@@ -951,6 +959,7 @@ int jent_entropy_init(unsigned int osr, unsigned int flags, void *hash_state,
 	struct rand_data *ec = p_ec;
 	int i, time_backwards = 0, ret = 0, ec_free = 0, delta_indx = 0;
 	unsigned int health_test_result;
+	bool init_in_progress = false;
 
 	if (!ec) {
 		ec = jent_entropy_collector_alloc(osr, flags, hash_state);
@@ -972,8 +981,11 @@ int jent_entropy_init(unsigned int osr, unsigned int flags, void *hash_state,
 		ec->health_failure &= (~JENT_APT_FAILURE);
 	}
 
+	if (!jcw_fips_enabled() && (flags == JENT_INIT))
+		init_in_progress = true;
+
 	/* priming of the ec->prev_time value */
-	jent_measure_jitter(ec, NULL);
+	jent_measure_jitter(ec, NULL, false);
 
 	/* We could perform statistical tests here, but the problem is
 	 * that we only have a few loop counts to do testing. These
@@ -1005,7 +1017,7 @@ int jent_entropy_init(unsigned int osr, unsigned int flags, void *hash_state,
 		__u64 start_time = 0, end_time = 0, delta = 0;
 
 		/* Invoke core entropy collection logic */
-		jent_measure_jitter(ec, &delta);
+		jent_measure_jitter(ec, &delta, init_in_progress);
 		end_time = ec->prev_time;
 		start_time = ec->prev_time - delta;
 
diff --git a/jitterentropy.h b/jitterentropy.h
index 95f4ce6..a56e17e 100644
--- a/jitterentropy.h
+++ b/jitterentropy.h
@@ -11,6 +11,9 @@
 
 #define JENT_CONF_DISABLE_LOOP_SHUFFLE
 
+/* Declare ASCII Value of Init to be used during initialization */
+#define JENT_INIT 0x494E4950
+
 extern void jent_get_nstime(__u64 *out);
 extern int jent_hash_time(void *hash_state, __u64 time, u8 *addtl,
 			  unsigned int addtl_len, __u64 hash_loop_cnt,
-- 
2.35.6

